internal/resource/resource_plugin_test.go

```diff
@@ -183,6 +191,194 @@ func TestResource_Process_Apply(t *testing.T) {
+			instance: protos.GetNginxPlusInstance([]string{}),
+		},
+		{
+			name: "Test 3: Fail, OSS Instance",
+			message: &bus.Message{
+				Topic: bus.APIActionRequestTopic,
+				Data: protos.CreatAPIActionRequestNginxPlusGetHTTPServers("test_upstream",
+					protos.GetNginxOssInstance([]string{}).GetInstanceMeta().GetInstanceId()),
+			},
+			err: errors.New("failed to preform API action, instance is not NGINX Plus"),
+			upstreams: []client.UpstreamServer{
+				helpers.CreateNginxPlusUpstreamServer(t, 10, 2, 0, 1234, true, false,
+					false, "test_server", "", "", "", ""),
+			},
+			topic:    []string{bus.DataPlaneResponseTopic},
+			instance: protos.GetNginxOssInstance([]string{}),
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(tt *testing.T) {
+			fakeResourceService := &resourcefakes.FakeResourceServiceInterface{}
+			fakeResourceService.GetUpstreamsReturns(test.upstreams, test.err)
+			fakeResourceService.InstanceReturns(test.instance)
+
+			messagePipe := busfakes.NewFakeMessagePipe()
+
+			resourcePlugin := NewResource(types.AgentConfig())
+			resourcePlugin.resourceService = fakeResourceService
+
+			err := messagePipe.Register(2, []bus.Plugin{resourcePlugin})
+			require.NoError(t, err)
+
+			resourcePlugin.messagePipe = messagePipe
+
+			resourcePlugin.Process(ctx, test.message)
+
+			assert.Equal(t, test.topic[0], messagePipe.GetMessages()[0].Topic)
+
+			response, ok := messagePipe.GetMessages()[0].Data.(*mpi.DataPlaneResponse)
+			assert.True(tt, ok)
+
+			if test.err != nil {
+				assert.Equal(tt, test.err.Error(), response.GetCommandResponse().GetError())
+				assert.Equal(tt, mpi.CommandResponse_COMMAND_STATUS_FAILURE, response.GetCommandResponse().GetStatus())
+			} else {
+				assert.Empty(t, response.GetCommandResponse().GetError())
+				assert.Equal(tt, mpi.CommandResponse_COMMAND_STATUS_OK, response.GetCommandResponse().GetStatus())
+			}
+		})
+	}
+}
+
+func TestResource_Process_APIAction_UpdateHTTPUpstreams(t *testing.T) {
+	ctx := context.Background()
+	tests := []struct {
+		instance    *mpi.Instance
+		name        string
+		expectedLog string
+		message     *bus.Message
+		err         error
+		topic       []string
+		upstreams   []client.UpstreamServer
+	}{
+		{
+			name: "Test 1: Success, Update Http Upstream Servers",
+			message: &bus.Message{
+				Topic: bus.APIActionRequestTopic,
+				Data: protos.CreatAPIActionRequestNginxPlusUpdateHTTPServers("test_upstream",
+					protos.GetNginxPlusInstance([]string{}).GetInstanceMeta().GetInstanceId(), []*structpb.Struct{
+						{
+							Fields: map[string]*structpb.Value{
+								"max_cons":  structpb.NewNumberValue(8),
+								"max_fails": structpb.NewNumberValue(0),
+								"backup":    structpb.NewBoolValue(true),
+								"service":   structpb.NewStringValue("test_server"),
+							},
+						},
+					}),
+			},
+			err: nil,
+			upstreams: []client.UpstreamServer{
+				helpers.CreateNginxPlusUpstreamServer(t, 10, 2, 0, 1234, true, false,
```

Comment:
You seem to call `CreateNginxPlusUpstreamServer` with the same parameters each time. Would it be better to have the function return an upstream server with some defaults set so that you dont need to same in all these parameters?